test/model/src/include/model/core.h

```diff
@@ -0,0 +1,406 @@
+    }
+
+    /*
+     * kv_update::operator!= --
+     *     Compare to another instance.
+     */
+    inline bool
+    operator!=(const kv_update &other) const noexcept
+    {
+        return !(*this == other);
+    }
+
+    /*
+     * kv_update::operator< --
+     *     Compare to another instance.
+     */
+    inline bool
+    operator<(const kv_update &other) const noexcept
+    {
+        if (_timestamp != other._timestamp)
+            return (_timestamp < other._timestamp);
+        if (_value != other._value)
+            return (_value < other._value);
+        return (true);
+    }
+
+    /*
+     * kv_update::operator<= --
+     *     Compare to another instance.
+     */
+    inline bool
+    operator<=(const kv_update &other) const noexcept
+    {
+        return (*this < other || *this == other);
+    }
+
+    /*
+     * kv_update::operator> --
+     *     Compare to another instance.
+     */
+    inline bool
+    operator>(const kv_update &other) const noexcept
+    {
+        return !(*this <= other);
+    }
+
+    /*
+     * kv_update::operator>= --
+     *     Compare to another instance.
+     */
+    inline bool
+    operator>=(const kv_update &other) const noexcept
+    {
+        return !(*this < other);
+    }
+
+    /*
+     * kv_update::value --
+     *     Get the value.
+     */
+    inline const data_value &
+    value() const noexcept
+    {
+        return (_value);
+    }
+
+    /*
+     * kv_update::global --
+     *     Check if this is a globally-visible, non-timestamped update.
+     */
+    inline bool
+    global() const noexcept
+    {
+        return (_timestamp == 0);
+    }
+
+    /*
+     * kv_update::value --
+     *     Get the value.
+     */
+    inline uint64_t
+    timestamp() const noexcept
+    {
+        return (_timestamp);
+    }
+
+private:
+    uint64_t _timestamp;
+    data_value _value;
+};
+
+/*
+ * kv_item --
+ *     The value part of a key-value pair, together with its metadata and previous versions.
+ */
+class kv_item {
+
+public:
+    /*
+     * kv_item::kv_item --
+     *     Create a new instance.
+     */
+    inline kv_item() noexcept {}
+
+    /*
+     * kv_item::~kv_item --
+     *     Delete the instance.
+     */
+    ~kv_item();
+
+    /*
+     * kv_item::add_update --
+     *     Add an update.
+     */
+    int add_update(kv_update &&update, bool must_exist, bool overwrite);
+
+    /*
+     * kv_item::contains_any --
+     *     Check whether the table contains the given value. If there are multiple values associated
+     *     with the given timestamp, return true if any of them match.
+     */
+    bool contains_any(const data_value &value, uint64_t timestamp = UINT64_MAX);
+
+    /*
+     * kv_item::get --
+     *     Get the corresponding value.
+     */
+    const data_value &get(uint64_t timestamp = UINT64_MAX);
+
+private:
+    std::mutex _lock;
+    std::vector<kv_update *> _updates; /* sorted list of updates */
```

Comment:
Does it make sense to use a more insert friendly container like `list` or append friendly `deque`? It's unclear to me what the nature of the majority of inserts will look like. You could then potentially drop the pointers and utilize `emplace`. This simplifies the dtor, etc.

Reply:
Most inserts will come to the end of the list, but do not have to be all the way at the end. When we eventually implement retiring old updates, we will be removing them from the front and replacing them by a single update in the front. I chose `vector` because appending and search are relatively efficient—I don't think `list` can do binary search well? Let me know what you think.

Reply:
As `vector` is analogous you're going to need to move memory if you insert at the beginning (rather than replace) or resize in the case `size() > capacity()` for appending. No afaik `list` does not provide fast random iterators, though `deque` does so it is better suited to `binary_search`.
I am not expecting any follow up, and I will leave it to resolve as you see fit.

Reply:
I didn't realize that about `deque`. In that case, switching to `deque` makes sense, so I just did that. Thank you for bringing this up!